### ISSUE TITLE
Replace RSA verification function `GFp_rsa_public_decrypt` with Rust code

### ISSUE THREAD

**Issue by briansmith** [briansmith/ring#246]
After the ECC implementation is improved, RSA signing and RSA verification will be the only two remaining operations that use `BIGNUM`. In particular, these will be the only two operations that use the heap. We would like to enable RSA signature verification in environemnts where there is no heap available too.

We should just write a new modular exponentiation function that always uses square-and-multiply and which avoids using `BIGNUM`. Note that verification doesn't need to be constant time. We can use `bn_mod_mul` for the modular multiplication.

Typically, the public exponent is `0x10001`. Sometimes it is 3 (0b11). It is rarely any other value.  Note that the typical exponents `0x10001` and `0b11` have hamming weight of 2. When the hamming weight is 2, the optimal algorithm is the simplest one: square and multiply.

Even if the exponent has a large hamming weight, the largest hamming weight possible is 33, because the RSA verification logic already restricts the public exponent to a maximum of 33 bits. This maximum was also contributed to BoringSSL so that Google Chrome could field-test that a 33-bit maximum is web-compatible. 

`BN_window_bits_for_exponent_size` already uses the basic square-and-multiply algorithm for exponents that are less than 24 bits; i.e. almost all the time. It only uses window-based exponentiation for exponents of 24 bits or more, and even then it only uses 3-bit windows.

Thus, the difference in performance in the best case should be minimal, and the difference in the worst case, which is extremely rare, will be only a little bit worse.

A good way to approach this project is to:
- [ ] Replace `BN_mod_exp_mont` with Rust code takes its input and output parameters as `&[Limb]` instead of `BIGNUM` and which uses `bn_mod_mul` for both squaring and multiplying, using the simple square-and-multiple algorithm. `Limb` should be moved from the `ring::ec` submodule to a new `ring::bn` module.
- [ ] Replace `GFp_rsa_public_decrypt` with Rust code that uses `BN_MONT_CTX_new()`/`BN_MONT_CTX_set()`/`BN_MOD_CTX_free()` and the new modular exponentiation function.
- [ ] Create a new variable-time modular inversion function that operates on `[Limb]`s (or, if in C, `BN_ULONG[]`) instead of `BIGNUM`, which will replace the variable-time `BN_mod_inverse_ex`.
- [ ] Create a replacement for `BN_MONT_CTX_new()`/`BN_MONT_CTX_set()`/`BN_MOD_CTX_free()` that avoids `BIGNUM`, using the new modular inversion function.

Note that I already have some code to implement the varable-time modular inversion using `BN_ULONG[]` in C.

**Comment by ranweiler** [comment]
I'm going to pick this one up.

**Comment by briansmith** [comment]
Awesome. Note:

> * Create a new variable-time modular inversion function that operates on [Limb]s (or, if in C, BN_ULONG[]) instead of BIGNUM, which will replace the variable-time BN_mod_inverse_ex.
> *  Create a replacement for BN_MONT_CTX_new()/BN_MONT_CTX_set()/BN_MOD_CTX_free() that avoids BIGNUM, using the new modular inversion function.

I have already done these two steps but I haven't published the code yet. LMK if you are interested in doing the code review for the new inversion code (which is actually still in C, because it comes from my old GFp project).

**Comment by ranweiler** [comment]
> I have already done these two steps but I haven't published the code yet. LMK if you are interested in doing the code review for the new inversion code (which is actually still in C, because it comes from my old GFp project).

Oh great! I'd definitely like to review it.

**Comment by briansmith** [comment]
See https://boringssl-review.googlesource.com/#/c/9031 for my first draft of the inversion part. I will put something together for *ring* shortly. I'm currently studying how to compute RR = r<sup>2</sup> mod p, which seems to be an annoying problem. Currently I have:

```
  /* RR = (2^ri)^2 == 2^(ri*2) == 1 << (ri*2), which has its (ri*2)th bit set. */
  int ri = (BN_num_bits(mod) + (BN_BITS2 - 1)) / BN_BITS2 * BN_BITS2;
  BN_zero(&(mont->RR));
  if (!BN_set_bit(&mont->RR, BN_num_bits(mod))) {
    return 0;
  }
  size_t expected_iterations = (2 * ri) - BN_num_bits(mod);
  for (size_t i = 0; ; ++i) {
    if (BN_cmp(&mont->RR, mod) >= 0) {
      if (!BN_sub(&mont->RR, &mont->RR, mod)) {
        return 0;
      }
      assert(BN_cmp(&mont->RR, mod) < 0);
    }
    if (i == expected_iterations) {
      break;
    }
    if (!BN_lshift1(&mont->RR, &mont->RR)) {
      return 0;
    }
  }
```

The benefit of the above is that it can be converted to non-heap-using, non-`BIGNUM` code that I have sitting around. But, I'd like to avoid doing `expected_iterations` shifts and reductions, so I'm looking for an alternative.

**Comment by briansmith** [comment]
I suggest we split this into two separate issues: Moving `GFp_rsa_public_decrypt` to Rust and moving `GFp_BN_mod_exp_mont_vartime` to Rust. When I originally filed this I envisioned the work on RSA happening in a different order than what actually happened, and I anticipated that we would avoid using `BIGNUM` in the Rust code. But, now we have `PositiveInteger` and in particular `PositiveInteger::from_be_bytes` and it seems pretty easy to move `GFp_rsa_public_decrypt` to Rust. In contrast, `GFp_BN_mod_exp_mont_vartime` seems likely to be blocked on other work.

@ranweiler Are you still interested in working on this?

**Comment by ranweiler** [comment]
@briansmith, I sure am (I was planning on picking this back up this week). Splitting things up in the way you suggest makes sense to me— let's do it!

**Comment by briansmith** [comment]
@ranweiler I accidentally did this as part of oxidizing RSA signature verification. This should make the more interesting part, #366, easier, at least.

**Comment by briansmith** [comment]
This landed in #373.
